presto-parquet/src/main/java/com/facebook/presto/parquet/reader/ParquetColumnChunk.java

```diff
@@ -100,51 +92,99 @@ public PageReader readAllPages(Optional<InternalFileDecryptor> fileDecryptor, in
+            byte[] pageHeaderAAD = headerBlockDecryptor != null ?
+                    AesCipher.createModuleAAD(fileDecryptor.get().getFileAAD(), ModuleType.DictionaryPageHeader, rowGroupOrdinal, columnOrdinal, -1)
+                    : null;
+            PageHeader pageHeader = readPageHeader(headerBlockDecryptor, pageHeaderAAD);
             int uncompressedPageSize = pageHeader.getUncompressed_page_size();
             int compressedPageSize = pageHeader.getCompressed_page_size();
-            long firstRowIndex = -1;
-            switch (pageHeader.type) {
-                case DICTIONARY_PAGE:
-                    if (dictionaryPage != null) {
-                        throw new ParquetCorruptionException("%s has more than one dictionary page in column chunk", descriptor.getColumnDescriptor());
-                    }
-                    dictionaryPage = readDictionaryPage(pageHeader, uncompressedPageSize, compressedPageSize);
-                    break;
-                case DATA_PAGE:
-                    firstRowIndex = PageReader.getFirstRowIndex(dataPageCount, offsetIndex);
-                    valueCount += readDataPageV1(pageHeader, uncompressedPageSize, compressedPageSize, firstRowIndex, pages);
-                    dataPageCount = dataPageCount + 1;
-                    pageOrdinal = pageOrdinal + 1;
-                    break;
-                case DATA_PAGE_V2:
-                    firstRowIndex = PageReader.getFirstRowIndex(dataPageCount, offsetIndex);
-                    valueCount += readDataPageV2(pageHeader, uncompressedPageSize, compressedPageSize, firstRowIndex, pages);
-                    dataPageCount = dataPageCount + 1;
-                    pageOrdinal = pageOrdinal + 1;
-                    break;
-                default:
-                    stream.skipFully(compressedPageSize);
-                    break;
-            }
+            dictionaryPage = readDictionaryPage(pageHeader, uncompressedPageSize, compressedPageSize);
         }
-        byte[] fileAdditionalAuthenticationData = (fileDecryptor.isPresent()) ? fileDecryptor.get().getFileAAD() : null;
+
+        final Iterator<DataPage> dataPageIterator = getDataPageIterator(dataPageHeaderAdditionalAuthenticationData, headerBlockDecryptor, dictionaryPage);
+
+        byte[] fileAAD = fileDecryptor.map(InternalFileDecryptor::getFileAAD).orElse(null);
         Optional<BlockCipher.Decryptor> dataDecryptor = getDataDecryptor(columnDecryptionSetup);
-        return new PageReader(descriptor.getColumnChunkMetaData().getCodec(), pages, dictionaryPage, offsetIndex, dataDecryptor, fileAdditionalAuthenticationData, rowGroupOrdinal, columnOrdinal);
+
+        long totalValueCount = descriptor.getColumnChunkMetaData().getValueCount();
+        return new PageReader(descriptor.getColumnChunkMetaData().getCodec(), dataPageIterator, totalValueCount,
+                dictionaryPage, offsetIndex, dataDecryptor, fileAAD, rowGroupOrdinal, columnOrdinal);
```

Comment:
Would you please use the existing name `fileAdditionalAuthenticationData` for `fileAAD`?